added telegram publisher

dkotlyar opened the pull request.

dkotlyar commented:
# Publisher service

Consumes notification topic from other service, prepare and publish 
messages to selected social media sources.

## Build

```commandline
docker build -t volunteer-prb/podcasts-api-publisher .
```

## Running

Local
```commandline
export TELEGRAM_TOKEN=<put_bot_token_here>
cd src && python main.py
```

In docker 
```commandline
docker run -e TELEGRAM_TOKEN=<put_bot_token_here> \
  -e REDIS_HOST=<redis_host> \
  --name publisher --rm \
  volunteer-prb/podcasts-api-publisher
```

## Example 

```python
import json

from redis import Redis


envelope = dict(
    title='Podcasts API Example',
   description='Project for uploading videos from YouTube as audio podcasts to TG and Soundcloud',
   hashtags=['volunteer', 'prb', 'media'],
   publisher='Volunteer PRB',
   photo='https://file-examples.com/storage/feb04797b46286b5ea5f061/2017/10/file_example_JPG_500kB.jpg',
   audio='https://file-examples.com/storage/feb04797b46286b5ea5f061/2017/11/file_example_MP3_700KB.mp3',
   recipients=[
       dict(
           _type_ = 'telegram',
           channel_id=-1000000000000  # replace it by you channel id
       )
   ]
)

r = Redis()
r.publish('notification_topic', json.dumps(envelope))
```

dkotlyar commented:
import json
import uuid

from redis import Redis


envelope = dict(
    title='Podcasts API Example',
    description='Project for uploading videos from YouTube as audio podcasts to TG and Soundcloud',
    hashtags=['volunteer', 'prb', 'media'],
    publisher='Volunteer PRB',
    photo='https://transfer.sh/get/shqEBi/photo.jpg',
    audio='https://transfer.sh/get/vYyWP8/audio.mp3',
    recipients=[
        dict(
            _type_ = 'telegram',
            channel_id=-1000000000000  # replace it by you channel id
        )
    ]
)

r = Redis()
id = str(uuid.uuid4())
data = json.dumps(envelope)
print(id, data)
r.set(f'notification_topic_{id}', data)
r.rpush('notification_topic_list', id)

id = str(uuid.uuid4())
envelope['title'] = 'Podcasts Second API Example'
data = json.dumps(envelope)
print(id, data)
r.set(f'notification_topic_{id}', data)
r.rpush('notification_topic_list', id)

dkotlyar commented:
# Publisher service

Consumes notification topic from other service, prepare and publish messages to selected social media sources.

## Build

```commandline
docker build -t volunteer-prb/podcasts-api-publisher .
```

## Running

Local
```commandline
export TELEGRAM_TOKEN=<put_bot_token_here>
cd src && celery --app=publisher --broker=redis://localhost worker --loglevel=INFO
```

In docker 
```commandline
docker run -e TELEGRAM_TOKEN=<put_bot_token_here> \
  -e BROKER_URL=redis://<redis_host> \
  --name publisher --rm \
  volunteer-prb/podcasts-api-publisher
```

In docker compose (increase maximum audio size to 2GB)
```yaml
version: '3.8'
services:
  telegram-bot-api:
    image: aiogram/telegram-bot-api:latest
    environment:
      TELEGRAM_API_ID: <put_telegram_api_id>
      TELEGRAM_API_HASH: <put_telegram_api_hash>
    volumes:
      - ./telegram-bot-api-data:/var/lib/telegram-bot-api

  publisher:
    image: volunteer-prb/podcasts-api-publisher:latest
    environment:
      TELEGRAM_SERVER: http://telegram-bot-api:8081
      TELEGRAM_TOKEN: <put_bot_token_here>
      BROKER_URL: redis://<redis_host>
    depends_on:
      - telegram-bot-api
```

Run with scale factor 3

```commandline
docker-compose up --scale publisher=3
```

## Example

```python
from publisher import publish


envelope = dict(
    title='Podcasts API Example',
    description='Project for uploading videos from YouTube as audio podcasts to TG and Soundcloud',
    hashtags=['volunteer', 'prb', 'media'],
    publisher='Volunteer PRB',
    photo='https://transfer.sh/get/shqEBi/photo.jpg',
    audio='https://transfer.sh/get/vYyWP8/audio.mp3',
    recipients=[
        dict(
            _type_='telegram',
            channel_id=-1000000000000  # replace it by you channel id
        )
    ]
)

publish.delay(envelope)
```

andreinechaev commented on publisher/src/telegram/publisher.py:
```diff
@@ -0,0 +1,100 @@
+TG_MAX_AUDIO_THUMB_SIZE = 320
+
+
+def publish(envelope: Envelope):
```
Please consider more granular methods for future reusability.

andreinechaev commented on publisher/src/telegram/publisher.py:
```diff
@@ -0,0 +1,100 @@
+        audio = filename
+
+    # filter recipient list, select only telegram recipients
+    for recipient in [rec for rec in envelope.recipients if isinstance(rec, TelegramRecipient)]:
```
such blocking for-loops are good candidates for queuing mechanism. Consider using Celery

dkotlyar replied:
good remark, i'll do that

dkotlyar commented:
## Changes 

* Replaced blocking for-loops with queuing mechanism
* Added dataclass serializer
* Message splitting function moved out for future reusability
* Updated README, updated Example of usage

## Example
```python
from app.objects.Envelope import EnvelopeBody, Envelope
from app.objects.Recipient import TelegramRecipient
from app.publisher import publish


envelope = Envelope(
    body=EnvelopeBody(
        title='Podcasts API Example',
        description='Project for uploading videos from YouTube as audio podcasts to TG and Soundcloud',
        hashtags=['volunteer', 'prb', 'media'],
        publisher='Volunteer PRB',
        photo='https://transfer.sh/get/shqEBi/photo.jpg',
        audio='https://transfer.sh/get/vYyWP8/audio.mp3',
    ),
    recipients=[
        TelegramRecipient(
            channel_id=-1000000000000  # replace it by you channel id
        )
    ]
)

publish.delay(envelope)

```

eIGato commented on mediamanager/main.py:
```diff
@@ -0,0 +1,23 @@
+    timezone='Europe/Paris',
+    enable_utc=True,
+    result_persistent=True,
+    ignore_result=False
```
Add a trailing comma.

eIGato commented on publisher/app/telegram/telegram_bot_api.py:
```diff
@@ -0,0 +1,310 @@
+    MAX_AUDIO_THUMB_SIZE = 320
+
+    """
+    The Bot API is an HTTP-based interface created for developers keen on building bots for Telegram.
```
Split the line to fit in 79 (or 80) chars.

eIGato commented on publisher/app/telegram/telegram_bot_api.py:
```diff
@@ -0,0 +1,310 @@
+        if json is None:
+            json = dict()
+        args = '&'.join([f'{k}={urllib.parse.quote_plus(str(v))}' for k, v in self.__filter(**args).items()]) if args is not None else ''
+        response = requests.post(f'{self.__server}/bot{self.__token}/{method_name}?{args}',
```
No lemmings please. Use uniform indents and feed the lines right after opening parentheses.

dkotlyar commented on publisher/app/telegram/telegram_bot_api.py:
```diff
@@ -0,0 +1,310 @@
+        response = requests.post(f'{self.__server}/bot{self.__token}/{method_name}?{args}',
+                                 json=self.__filter(**json),
+                                 files=files)
+        if response.status_code == 200 and (json := response.json()) and \
```
How can I change long if condition without backslashes in the line end?

dimon222 commented on publisher/app/telegram/telegram_bot_api.py:
```diff
@@ -32,42 +43,68 @@ def request(self, method_name, json=None, files=None, args=None):
+            json=dict_filter(json),
+            files=files,
+        )
+        if response.status_code == 200 and (json := response.json()) is not None \
+                and 'ok' in json and json['ok'] is True and 'result' in json:
```
Perhaps we should just check the status code only?
```python
response.raise_for_status()
return json.get('result')
```

dkotlyar replied:
I'm not sure we should just check only status code, I rewrote this section with raise error contains Telegram description.

```python
try:
    if not response.json().get('ok'):
        raise TelegramBotError(response.json().get('description'))
    return response.json().get('result')
except json.decoder.JSONDecodeError:
    response.raise_for_status()
    raise TelegramBotError('Not JSON response')
```

dimon222 commented on publisher/app/telegram/telegram_bot_api.py:
```diff
@@ -32,42 +43,68 @@ def request(self, method_name, json=None, files=None, args=None):
-                                 files=files)
-        if response.status_code == 200 and (json := response.json()) and \
-                'ok' in json and json['ok'] is True and 'result' in json:
+        args = '&'.join([
+            f'{k}={quote_plus(str(v))}' for k, v in dict_filter(args).items()
+        ])
+        response = requests.post(
+            url=f'{self.__server}/bot{self.__token}/{method_name}?{args}',
+            json=dict_filter(json),
+            files=files,
+        )
```
You don't need to do this work with manually reconstructing GET params, there's "good practice" sauce out of the box. Keep `args` as dict and then pass it as `data` argument in function.
```python
response = requests.post(
            url=f'{self.__server}/bot{self.__token}/{method_name},
            json=dict_filter(json),
            data=args,
            files=files,
        )

dimon222 commented:
Maybe its a minor thing, but since there's relatively big adjustment done in this PR and my "annoying" side recommendation would be to avoid using conflicting "standard lib" naming for any of arguments.
That means using variable like `json` should be discouraged as there's library `json` that might cause serious confusion. Instead we could consider using `json_body`, `body` or such.

dimon222 commented on publisher/app/telegram/publisher.py:
```diff
@@ -0,0 +1,84 @@
+
+    # download photo if url passed, make thumb
+    photo = envelope.photo
+    if photo is not None:
```
Perhaps can be simplified to more pythonic:
```python
if photo:
```

dimon222 commented on publisher/app/telegram/telegram_bot_api.py:
```diff
@@ -0,0 +1,431 @@
+            if (len(buffer) + len(partial) + len(splitter)) < max_len():
+                buffer += splitter + partial
+            else:
+                if len(buffer.strip()) > 0:
```
More pythonic:
```python
if buffer.strip():
```

andreinechaev reviewed (commented):
Please rebase on master.
Please consider simple testing.
Besides those remarks - it's an exemplary PR with documentation and clear execution. Good work!

dkotlyar commented:
> Please rebase on master. Please consider simple testing. Besides those remarks - it's an exemplary PR with documentation and clear execution. Good work!

Something went wrong while rebase, I hope it's okay, I could made only force-push.

About simple testing I wrote an example [above](https://github.com/volunteer-prb/podcasts-api/pull/2#issuecomment-1135104721) to run and publish to Telegram. What do you recommend to add to my PR? Maybe I can add test function to check Telegram answer, but I'm not sure. 

Thank you for your remarks.

andreinechaev commented on publisher/app/telegram/telegram_bot_api.py:
```diff
@@ -0,0 +1,443 @@
+    pass
+
+
+class TelegramBot:
```
Looks like we're reinventing the wheel
Why not using the python lib https://github.com/python-telegram-bot/python-telegram-bot ?

andreinechaev replied:
nvm. The solution uses local sever to upload larger files.
